state-chain/pallets/cf-swapping/src/mock.rs

```diff
@@ -91,6 +92,20 @@ impl MockSwappingApi {
 	fn swaps_should_fail() -> bool {
 		SWAPS_SHOULD_FAIL.with(|cell| cell.get())
 	}
+
+	pub fn add_liquidity(asset: Asset, amount: AssetAmount) {
+		let liquidity = Liquidity::get()
```

Comment:
I was thinking there might be a simpler way to simulate this. We don't need to add amounts etc, we could just configure the mock to behave *as if* there is no liquidity in a certain pool (ie. instead of a liquidity amount we just have a bool).

This would avoid having to calculate/reason about how much input is required at current prices to make the test fail etc.

Reply:
I almost ended up making this change, but then ran into the `price_impact_removes_one_swap` test where we actually want low liquidity, but enough for *one* swap to test the retrying works (and even the swaps from the failing pool/side can be retired). Because retrying is done entirely within a single call to on_finalize there is no way for us to change liquidity between iteration as we remove swaps, so we need a bit more control in the mock's liquidity than to just say "all swaps in the pool should fail".

Also, I thought it was nice that we could check the resulting liquidity in presence of rollbacks (e.g. in `storage_state_rolls_back_on_fok_violation`), which I thought was the whole point of marking `Liquidity` as `storage`.